tests/check_framework/test_files.py

```diff
@@ -0,0 +1,36 @@
+from pathlib import Path
+
+from django.core.checks import Error
+from django.core.checks.files import E001, check_setting_file_upload_temp_dir
+from django.test import SimpleTestCase
+from django.test.utils import override_settings
+
+TEST_PATH = str(Path(__file__).resolve().parent)
+
+
+class FilesCheckTests(SimpleTestCase):
+    @override_settings(FILE_UPLOAD_TEMP_DIR=None)
+    def test_file_upload_temp_dir__none(self):
+        self.assertEqual(check_setting_file_upload_temp_dir(None), [])
+
+    @override_settings(FILE_UPLOAD_TEMP_DIR="non-existing-folder")
+    def test_file_upload_temp_dir__non_existing_str(self):
+        self.assertEqual(
+            check_setting_file_upload_temp_dir(None),
+            [Error(E001.msg.format("non-existing-folder"), id=E001.id)],
+        )
+
+    @override_settings(FILE_UPLOAD_TEMP_DIR=Path("non-existing-folder"))
+    def test_file_upload_temp_dir__non_existing_path(self):
+        self.assertEqual(
+            check_setting_file_upload_temp_dir(None),
+            [Error(E001.msg.format("non-existing-folder"), id=E001.id)],
+        )
+
+    @override_settings(FILE_UPLOAD_TEMP_DIR=TEST_PATH)
+    def test_file_upload_temp_dir__existing_str(self):
+        self.assertEqual(check_setting_file_upload_temp_dir(None), [])
+
+    @override_settings(FILE_UPLOAD_TEMP_DIR=Path(TEST_PATH))
+    def test_file_upload_temp_dir__existing_path(self):
+        self.assertEqual(check_setting_file_upload_temp_dir(None), [])
```

Comment:
We can use a `subTest()`:
```suggestion
    def test_file_upload_temp_dir(self):
        tests = [
            None,
            '',
            Path.cwd(),
            str(Path.cwd()),
        ]
        for setting in tests:
            with self.subTest(setting):
                with self.settings(FILE_UPLOAD_TEMP_DIR=setting):
                    self.assertEqual(check_setting_file_upload_temp_dir(None), [])

    def test_file_upload_temp_dir_nonexisten(self):
        for setting in ['nonexistent', Path('nonexistent')]:
            with self.subTest(setting):
                with self.settings(FILE_UPLOAD_TEMP_DIR=setting):
                    self.assertEqual(check_setting_file_upload_temp_dir(None), [
                        Error(
                            "The FILE_UPLOAD_TEMP_DIR setting refers to the "
                            "nonexistent directory 'nonexistent'.",
                            id='files.E001',
                        ),
                    ])
```